README.md
# [PSO2NA][PSO2] UQ Parser

## Overview

This project scrapes from the [Urgent Quests main page](https://pso2.com/news/urgent-quests) to parse schedules into a database for easier reading. It works by going through the UQ main page and going through each schedule. If a schedule is not found in the database, meaning no entries in the database have the schedule's URL, the schedule is subsequently scraped for Urgent Quests.

Additionally, webhook and RSS functionality are included. They should be run separately from the main file and only after running the main script.

**Note that RSS functionality may be sunset and deprecated in the future.**

## Usage

Install dependencies, and run [main.py](main.py). Ideally, the script should be run once a day at midnight server time (i.e. `America/Los_Angeles`). **Do not create a `main.yaml`!** If you are running the project for the first time, let the project handle it.

Once you have at least the main script once, you can run [webhook.py](webhook.py) or [rss.py](rss.py). Like the main script, ideally these should run on a schedule, preferably every half hour (`:00` and `:30`).

## [Requirements](requirements.txt)

This code is designed around the following:

- Python 3.7+
    - `pendulum`
    - `pyyaml`
    - `requests`

## Disclaimer

This project is not affiliated with or endorsed by *[PSO2][PSO2]*, *[SEGA][SEGA]*, or *[Microsoft][MICROSOFT]*. See [LICENSE](LICENSE) for more detail.

[PSO2]: https://pso2.com/
[SEGA]: https://www.sega.com/
[MICROSOFT]: https://www.microsoft.com/


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [1.1.7] - 2020-09-03
### Fixed
- #2 - The schedule for 2020-09-01 introduces a new column in every time table for times in GMT; to address this, the code now counts how many columns contain "Time (.*)" (regex) to accurately offset the cell checking.

## [1.1.6] - 2020-07-22
### Fixed
- In [uq.py](uq.py), `black` as a color background no longer causes `KeyError` while comparing Euclidean color distances.

## [1.1.5] - 2020-07-18
### Changed
- In [uq.py](uq.py):
    - `get_uq_from_cell()` no longer calls `get_closest_color()` when it doesn't find a match; instead, it will raise the new `MismatchedColor` exception.
    - `get_closest_color()` now has an optional parameter `is_uq` (default: `True`) to filter events by UQ or other (concerts, etc.). Related to #1.
    - `get_closest_color()` is now called by `Schedule.parse()`; `Schedule.parse()` will check whether a prior event is listed in a schedule before determining whether an event is a UQ or something else.

### Fixed
- #1 - In [uq.py](uq.py), `get_closest_color()` now properly identifies the appropriate "closest" color. For example, if an event was intended to be a concert but its closest color was an urgent quest, the concert should now be picked instead. For more context, look up issue #1 on GitHub.

## [1.1.4] - 2020-07-07
### Added
- Added a blacklist. Prior to this change, the scraper would keep opening https://pso2.com/news/urgent-quests/about because the page contained no UQs in the database. The scraper only knows not to scrape if the database contains at least one entry with the URL. The [blacklist](blacklist.yaml) is not meant to be configured separately; please commit changes and/or file a pull request.

### Changed
- Changed the repo name to **PSO2NA UQ Parser**.
- Readme now actually contains information!

## [1.1.3] - 2020-06-19
### Changed
- In [uq.py](uq.py):
    - Renamed `hexcolor` in `get_uq_from_cell` to `color`, as the color can be a hard-coded string (e.g. `red`) or a hex representation.
    - Renamed `get_colors_from_table()` to `get_colors_from_key()`, as `key` is a more accurate descriptor of the HTML table.
    - Instead of relying on hard-coded colors, a new function `get_closest_color()` was added: this should get the correct UQ if colors don't match exactly between the schedule and the color key.

## [1.1.2] - 2020-06-10
### Changed
- In [uq.py](uq.py):
    - `parse_date()` now considers cases where a schedule contains an event in the past year (e.g. December) posted in the new year (e.g. January).
    - Combine `MainPage.delete_old()` with `MainPage.parse()` since `delete_old()` required `parse()` anyway.
- In [webhook.py](webhook.py):
    - Instead of only checking the direct next event, limit the results based on `LAST` known index.

### Fixed
- Added a specific check in `get_uq_from_cell()` for a hard-coded mismatch in colors (key and schedule colors differ).
- `webhook.execute_webhook()` should now update the `LAST` selected event.

## [1.1.1] - 2020-06-05
### Changed
- Instead of stopping parsing in `MainPage.parse()`, continue to parse, only ignoring existing entries. (They will be skipped.)

### Fixed
- Fixed incorrect reference to `UQMainPage` (see 1.0.0) in [main.py](main.py).
- Fixed `self.schedules` not being `self.schedules.items()` in [uq.py](uq.py).

## [1.1.0] - 2020-06-04
### Added
- Added a new function `MainPage.delete_old()` in [uq.py](uq.py) that deletes schedules (and their associated UQs) if the schedule is no longer found on the main page. This is run in [main.py](main.py). Note that if `MainPage.parse()` is not called prior, `delete_old()` *will* error.

### Changed
- Uncommented a `+ 1` to `parse_date()` in [uq.py](uq.py) for years. The +1 is meant for incrementing year for future events in the new year. e.g. the script is run in December and found a new schedule in January. The year will be bumped up based on months (`12` and  `1` respectively). If results have not been cached or updated, comment out the `+ 1`.
- Webhooks are now generic, no longer tied to *Discord*, as long as the method is `POST`. Consequently, this means you should use the full URL in the configuration, not just the ID.

## [1.0.0] - 2020-05-31
### Added
- Added *[Discord][DISCORD]* [webhook](webhook.py) support. Set this up by copying [webhook.yaml.example](webhook.yaml.example) to `webhook.yaml` and fill out the ID of the webhook. This project is not affiliated or endorsed by *[Discord][DISCORD]*.
- Added a check that ensures scraping only new schedules by getting the most recent entry and getting its schedule title.

### Changed
- In [uq.py](uq.py):
    - `UQMainPage` and `UQSchedule` were renamed `MainPage` and `Schedule`, respectively. The `UQ` is already implied with the module name.
- In [config.py](config.py):
    - The entire record of UQs is retrieved from the database. Formerly, `results` were used in [uq.py](uq.py) and [rss.py](rss.py). Currently, `rss.py` has not been updated. This point relates to the 2nd bullet point under "Added".

## [0.1.1] - 2020-05-25
### Fixed
- ~~Removed GUID per RSS entry, as the feed won't validate.~~ Just kidding, changed the value because RSS readers can't handle this.
- Added missing dependency `feedgen`.

## [0.1.0] - 2020-05-25
### Added
- Initial version

[DISCORD]: https://discord.com


rss.py
from datetime import timedelta

import pendulum
from feedgen.feed import FeedGenerator

import config


AUTHOR = {'name': 'SEGA'}

LINK = 

class UQRSS:
    """RSS feed for Urgent Quests."""

    PRIOR_MINS = timedelta(minutes=30)
    EVENT_TIME = str(config.NOW + PRIOR_MINS)

    def __init__(self) -> None:
        """Initialize the RSS generation first by initializing feedgen."""
        self.fg = FeedGenerator()
        self.fg.title('PSO2 News: Urgent Quests')
        self.fg.author(AUTHOR)
        self.fg.description('Phantasy Star Online 2 News tracker for the West')
        self.fg.link(
            href='https://pso2.com/news/urgent-quests',
            rel='alternate',
            )
        self.fg.link(
            href=LINK,
            rel='self',
            )
        self.fg.language('en-US')

    def generate_feed(self) -> None:
        """Generate a feed by going through the database."""
        config.CURSOR.execute('SELECT * FROM UQ')
        results = sorted(config.CURSOR.fetchall(), reverse=True)
        start = 0
        for n, (dt_str, uq, title, url) in enumerate(results):
            if dt_str > self.EVENT_TIME:
                continue
            elif start == 0:
                start = n
            elif n - start > 10:
                return
            entry = self.fg.add_entry()
            entry.title(uq)
            entry.author(AUTHOR)
            #entry.description(uq)
            entry.link(href=url)
            entry.guid(f'{dt_str}/{uq}')
            entry.pubDate(pendulum.parse(dt_str))

    def write_feed(self) -> None:
        """Write out the feed after generating entries."""
        if len(self.fg.entry()) > 0:
            self.fg.rss_file('uq.xml')
        else:
            config.LOGGER.error('No entries were generated!')


if __name__ == '__main__':
    uq = UQRSS()
    uq.generate_feed()
    uq.write_feed()



config.py
import logging
import logging.handlers
import sqlite3

import pendulum
import yaml


LOGGER = logging.getLogger('pso2_news')
LOGGER.setLevel(logging.DEBUG)

FH = logging.handlers.RotatingFileHandler(
    'news.log',
    maxBytes=4096,
    backupCount=5,
    )
FH.setLevel(logging.DEBUG)

CH = logging.StreamHandler()
CH.setLevel(logging.INFO)

FH.setFormatter(
    logging.Formatter(
        '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
        )
    )
CH.setFormatter(
    logging.Formatter(
        '%(levelname)s - %(message)s'
        )
    )

LOGGER.addHandler(FH)
LOGGER.addHandler(CH)

TODAY = pendulum.today()
NOW = pendulum.now()


DB = sqlite3.Connection('news.db')
CURSOR = DB.cursor()

SCHEMA = {
    'UQ': '(DATE TEXT UNIQUE, NAME TEXT, TITLE TEXT, URL TEXT)'
    }

for table, schema in SCHEMA.items():
    CURSOR.execute('CREATE TABLE IF NOT EXISTS {0} {1}'.format(table, schema))
    DB.commit()


CURSOR.execute('SELECT * FROM UQ')
RESULTS = sorted(CURSOR.fetchall(), reverse=True)


try:
    with open('main.yaml', 'r') as f:
        yaml.safe_load(f)['FIRST_RUN']
        FIRST_RUN = False
except (FileNotFoundError, KeyError, TypeError):
    FIRST_RUN = True
    LOGGER.info(
        'This is a first run. '
        'A first run will make subtle changes to the year of UQs.'
        )

with open('blacklist.yaml', 'r') as f:
    BLACKLIST = yaml.safe_load(f)

try:
    UQ_BLACKLIST = BLACKLIST['uq']
except (KeyError, TypeError):
    UQ_BLACKLIST = []
    LOGGER.warning('The blacklist is malformed. Please download a new copy.')


def write_main() -> None:
    """Write main configuration yaml.

    Currently, main.yaml should only include a "FIRST_RUN" element.

    """
    if FIRST_RUN:
        with open('main.yaml', 'w') as f:
            yaml.safe_dump({'FIRST_RUN': False})


webhook.py
from datetime import timedelta
from typing import Dict

import pendulum
import requests
import yaml

import config


NEXT = config.NOW + timedelta(minutes=30)

MESSAGE = """
Time: **{0}** (less than 30 minutes)

ᴿᵉᵃᵈ ᵗʰᵉ [ˢᶜʰᵉᵈᵘˡᵉ]({1})
"""

with open('webhook.yaml', 'r') as f:
    CONF = yaml.safe_load(f)
    ID = CONF['ID']
    try:
        LAST = CONF['LAST']
    except KeyError:
        LAST = None


def execute_webhook(dt: pendulum.datetime, uq: str, title: str) -> None:
    """Execute webhook given a UQ's name, datetime, and title of the
    event page it belongs.

    Args:
        dt (pendulum.datetime): the datetime of the UQ
        uq (str): name of the UQ
        title (str): title of the page that had the UQ on schedule

    """
    payload = {
        "embeds": [
            {
                "title": f"**{uq}**",
                "description": MESSAGE.format(
                    dt.to_day_datetime_string(),
                    title,
                    )
                }
            ]
        }
    response = requests.post(ID, json=payload)
    out = {
        'ID': ID,
        'LAST': str(dt),
        }
    with open('webhook.yaml', 'w') as f:
        yaml.safe_dump(out, stream=f)

    config.LOGGER.info(f'Executed webhook @ {config.NOW}: {response}')
    config.LOGGER.info(f'UQ: {uq}, DT: {dt}')



def search_events() -> None:
    """Search events by going through the database, finding one that
    will happen between now (to be run at :00 and :30) and 30 minutes
    later.

    """
    if LAST:
        dt_strs = [dt_str for dt_str, uq, title, url in config.RESULTS]
        index = dt_strs.index(LAST)
        # dt_str, uq, title, url = config.RESULTS[index - 1]
        # dt = pendulum.parse(dt_str)
        # if config.NOW <= dt <= NEXT:
        #     execute_webhook(dt, uq, url)
        #     return
    else:
        index = None
    for dt_str, uq, title, url in config.RESULTS[:index]:
        dt = pendulum.parse(dt_str)
        # In reverse chronological order, some events may be ahead.
        # Those events should be ignored.
        if dt > NEXT:
            continue
        # Likewise, some events will be behind. If an event hasn't
        # been found in range (they cannot have collisions),
        # stop looking.
        elif dt < config.NOW:
            return
        else:
            execute_webhook(dt, uq, url)
            return


if __name__ == '__main__':
    search_events()


uq.py
import re
import sqlite3
from datetime import timedelta
from math import ceil
from time import sleep
from typing import Callable, Dict, List, Tuple, Union

import pendulum
import requests
from bs4 import BeautifulSoup, Tag
from more_itertools import grouper

import config


EXAMPLE_MAIN = 'example-urgent_quests.html'
EXAMPLE_SCHEDS = [
    '2020-02',
    '2020-03',
    '2020-05_1',
    '2020-05_3',
    'about',
    ]


MONTHS = [
    'January', 'February', 'March', 'April', 'May', 'June',
    'July', 'August', 'September', 'October', 'November', 'December'
    ]

N = re.compile(r'[0-9]+')
RGB = re.compile(r'([0-9]+), ([0-9]+), ([0-9]+)')
TIME = re.compile(r'[0-9]{1,2}:[0-9][0-9]([ap]m)?', re.I)
AMPM = re.compile(r'[ap]m', re.I)
DT = re.compile(r'Time (.*)')
PSDT = re.compile(r'P[^T]*T')

NOT_UQ = [
    "Server Maintenance (Users won't be able to log in)",
    "Servers Open (Start of Extended Period)",
    "Server Shutdown (End of the Closed Beta Test)",
    ]

# hex string: tuple of int
KEY_COLORS = {}


def parse_date(month: int, day: int) -> Tuple[int, int, int]:
    """Parse a date given month and day only and convert to
    a tuple.

    Args:
        month (int): 1-index month value (e.g. 1 for January)
        day (int): a day of the month

    Returns:
        Tuple[int, int, int]: (year, month, day)

    """
    if month < config.TODAY.month:
        # Note that if you have not yet recorded/cached the current
        # records, you should comment out the +1. The +1 is only
        # meant to increment for future events that happen in
        # the new year.
        year = config.TODAY.year + 1
    elif month - config.TODAY.month > 1:
        # I realized that on June 10th, 2020, the schedule for UQs was
        # posted June 10th but included June 9th (which had passed).
        # There is a distinct possibility that this will happen again,
        # when the schedule is posted on New Year's Day (around there)
        # and includes a day for December. Because events are only
        # at most a month away in the future, we should check whether
        # the difference in months is greater than 1.
        # e.g. 12 - 1 > 1 to represent December of previous year and
        # January of the current year
        year = config.TODAY.year - 1
    else:
        year = config.TODAY.year
    return year, month, day


def parse_special_date(date: str) -> Callable[[int, int], Tuple[int, int, int]]:
    """Parse a date like "February 8th" into pendulum.datetime.

    Args:
        date (str): a date like "1/1" representing January 1

    Returns:
        Callable[[int, int], Tuple[int, int, int]]: return `parse_date()`

    """
    month, day = date.split(' ')[:2]
    month = MONTHS.index(month) + 1
    day = int(N.search(day).group(0))
    return parse_date(month, day)


def convert_time(time: str, ampm: str) -> Tuple[int, int]:
    """Convert time given "HH:MM" to 24h format.

    Args:
        time (str): a time like "12:00" without ampm
        ampm (str): either "am" or "pm"

    Returns:
        Tuple[int, int]: (hour, minute) in 24h time format

    """
    hour, minute = [int(n) for n in time.split(':')]
    hour %= 12
    if ampm == 'pm':
        hour += 12
    return hour, minute


def parse_time(time: str) -> Callable[[str, str], Tuple[int, int]]:
    """Parse time like "1:00 AM" and convert.

    Args:
        time (str): a time like "12:00am"

    Returns:
        Callable[[str, str], Tuple[int, int]]: return `convert_time()`

    """
    time, ampm = time.split(' ')
    return convert_time(time, ampm.lower())


def parse_time_range(time: str) -> Callable[[str, str], Tuple[int, int]]:
    """Parse time like "0:00 - 0:30am" and convert.

    Args:
        time (str): a time range like "12:00am - 1:00pm"

    Returns:
        Callable[[str, str], Tuple[int, int]]: return `convert_time()`

    """
    start, end = [TIME.search(t).group(0) for t in time.split(' – ')]
    return convert_time(start, AMPM.search(end).group(0))


def is_not_uq(uq: str) -> bool:
    """Is the UQ actually not a UQ? Filter by name here.

    Args:
        uq (str): name of the UQ

    Returns:
        bool: whether the UQ is actually a UQ

    """
    return uq in NOT_UQ


def get_uq_from_cell(cell: Tag, colors: Dict[str, str]) -> str:
    """Get a UQ from a cell given the cell's color.

    Args:
        cell (Tag): a cell in a table representing a UQ; its attributes
            include the color which is extracted
        colors (Dict[str, str]): a dictionary mapping colors from a key
            to UQs; cells must either match a color here or be ignored

    Returns:
        str: if valid, the UQ name associated with a color

    Raises:
        MismatchedColor: if the specified color wasn't found, raise the
            exception to pass the color, check in `get_closest_color`.

    """
    for attr in cell['style'].split(';'):
        if not attr:
            continue
        a, value = attr.split(':')
        if a == 'background':
            try:
                color = value.strip()
                return colors[color]
            except KeyError:
                raise MismatchedColor(color)


def get_hex_color_from_cell(cell: Tag) -> str:
    """Get a HEX color from the cell background, or if the background
    color is "red", just "red".

    Args:
        cell (Tag): a <td> element containing only a color

    Returns:
        str: a hex format string representing a color

    """
    for attr in cell['style'].split(';'):
        if not attr:
            continue
        a, value = attr.split(':')
        if a == 'background':
            match = RGB.search(value)
            if match:
                rgb_int = [int(n) for n in match.group(1, 2, 3)]
                rgb = '#' + ''.join(
                    [hex(n)[2:] for n in rgb_int]
                    ).upper()
                if rgb not in KEY_COLORS:
                    KEY_COLORS[rgb] = rgb_int
                return rgb
            else:
                # This is horrifying. Hard-coded colors.
                color = value.strip()
                if color == 'black':
                    color = '#000000'
                    if color not in KEY_COLORS:
                        KEY_COLORS[color] = (0, 0, 0)
                return color


def get_colors_from_key(table: Tag) -> Dict[str, str]:
    """Map a color from RGB to HEX to its UQ.

    Args:
        table (Tag): represents a HTML table of a color key

    Returns:
        Dict[str, str]: a dictionary mapping colors to UQs

    """
    colors = {}
    for row in table.find_all('tr'):
        col_color, col_uq = row.find_all('td')
        color = get_hex_color_from_cell(col_color)
        colors[color] = col_uq.text.replace('\xa0', ' ')

    return colors


def get_closest_color(
    color: str, colors: Dict[str, str], is_uq: bool = True
    ) -> Union[str, None]:
    """Get the closest color representation. Only works for hex colors.

        e.g. '#4F2CD0' matches with '#341D8B'

    This is because the UQ schedule appears to use hand-picked colors,
    so some inconsistencies are present.

    Original comment:
        Used for example-urgent_quest-2020-06_1.html;
        Urgent Quest:
            The Manifested Planetbreaker &
            The Chant to Cleanse the Calamity (60 minutes)
        The key color is #341D8B while the schedule color is
        #4F2CD0.

    Args:
        color (str): a color representation; should be in hex
        colors (Dict[str, str]): a dictionary mapping colors from a key
            to UQs; cells must either match a color here or be ignored
        is_uq (bool, optional): whether to limit to UQs xor concerts;
            defaults to True

    Returns:
        str: if valid, the UQ name associated with a color
        None: if no colors were matched; probably an empty cell

    """
    # Not a hex color; abort comparison
    if not color.startswith('#'):
        return
    else:
        color = color.replace('#', '')

    try:
        rgb_int = [int(f'{x}{y}', base=16) for x, y in grouper(color, 2)]
    except ValueError:
        config.LOGGER.warning(f'{color} could not be converted via int()')
        return

    # Maximum color Euclidean distance between black (#000000) and
    # white (#FFFFFF)
    distance = 3 * 255**2
    closest = None

    for c, uq in colors.items():
        if uq.startswith('Urgent Quest:') ^ is_uq:
            continue
        # Get Euclidean distance of the colors; using square value
        d = sum([(c1 - c2)**2 for c1, c2 in zip(rgb_int, KEY_COLORS[c])])
        if d < distance:
            distance = d
            closest = uq

    return closest


class MismatchedColor(ValueError):
    """A mismatched color was found."""
    def __init__(self, color: str) -> None:
        self.color = color


class Schedule:
    """Represents a schedule page for Urgent Quests."""

    def __init__(
        self, url_or_file: str, *, title: str = None, is_url: bool = True
        ) -> None:
        """Initialize the schedule parser with a URL or local file.

        Args:
            url_or_file (str): the URL or local file of a UQ schedule
            title (str, optional): the title of the schedule;
                defaults to None
            is_url (bool, optional): is url_or_file a URL?
                defaults to None

        """
        config.LOGGER.info(f'Initializing UQ Schedule @ {url_or_file}')
        self.title = title
        self.is_url = is_url
        self.url = url_or_file
        if is_url:
            page = requests.get(url_or_file)
            self.soup = BeautifulSoup(page.text, 'html.parser')
        else:
            with open(f'example-urgent_quest-{schedule}.html', 'r') as example:
                self.soup = BeautifulSoup(example, 'html.parser')

    def parse(self) -> None:
        """Parse the page and convert into database entries."""
        self.schedule = {}
        tables = self.soup.find('div', 'emergency cms')
        for table_a, table_b in grouper(tables.find_all('table'), 2):
            rows = table_a.find_all('tr')
            cols = rows[0].find_all('td')
            width = float(cols[1]['width'].strip('%'))
            if len(cols) == 1:
                rows.pop(0)
                cols = rows[0].find_all('td')
            # Special case: example-urgent_quest-2020-02.html
            if len(cols) == 2:
                # Because 2020-02 does not have a color code per table,
                # iterate over both tables instead.
                self.parse_only_tables([table_a, table_b])
            else:
                dates = [
                    parse_date(*[int(n) for n in col.text.split('/')])
                    for col in cols[1:]
                    ]
                timezones = 0
                pacific = 0
                for i, cell in enumerate(rows[2].find_all('td')):
                    if DT.search(cell.text):
                        timezones += 1
                        if PSDT.search(cell.text):
                            pacific = i
                # Skip row 2 (days of the week) and row 3 ("Time (PDT)").
                color_map = get_colors_from_key(table_b)
                for row in rows[3:]:
                    widths = 0
                    try:
                        time = parse_time(row.find_all('td')[pacific].text)
                    except ValueError:
                         # Some tables have empty rows under the table. Why.
                         continue
                    for cell in row.find_all('td')[timezones:]:
                        widths += float(cell['width'].strip('%'))
                        dt = pendulum.datetime(
                            *(dates[ceil(widths/width) - 1] + time),
                            tz='America/Los_Angeles'
                            )
                        try:
                            uq = get_uq_from_cell(cell, color_map)
                        except MismatchedColor as e:
                            if dt.minute == 30:
                                dt0 = dt + timedelta(minutes=-30)
                                # If an entry in the schedule exists 30min
                                # prior to this entry, it's a 60min UQ.
                                is_uq = dt0 in self.schedule
                            else:
                                # Only UQs start at the top of the hour.
                                is_uq = True
                            if e.color == 'black':
                                e.color = '#000000'
                            uq = get_closest_color(e.color, color_map, is_uq)

                        if not uq:
                            continue
                        self.schedule[dt] = uq

        self.write_to_db()

    def parse_only_tables(self, tables: List[Tag]) -> None:
        """Parse a table, ignoring any color code. Used in 2020-02.

        Args:
            tables (List[Tag]): a list of html tables; tables here are
                2-columns wide, with time ranges in column 1 and UQ name
                in column 2

        """
        for table in tables:
            rows = table.find_all('tr')
            cols = rows[0].find_all('td')
            # Special case: 2nd table of example-urgent_quest-2020-02.html
            if len(cols) == 1:
                rows.pop(0)
                cols = rows[0].find_all('td')
            year, month, day = parse_special_date(
                table.previous_sibling.text
                )
            for row in rows[1:]:
                time, uq = [tag.text.strip() for tag in row.find_all('td')]
                # Special case:
                #   2nd table of example-urgent_quest-2020-02.html
                if time == 'Time (PST)':
                    continue
                if is_not_uq(uq):
                    continue
                hour, minute = parse_time_range(time)
                dt = pendulum.datetime(year, month, day, hour, minute)
                self.schedule[dt] = uq

    def write_to_db(self) -> None:
        """Write the schedule to DB."""
        records_in = 0
        config.CURSOR.execute('SELECT DATE FROM UQ')
        dates = config.CURSOR.fetchall()
        for date, uq in self.schedule.items():
            try:
                config.CURSOR.execute(
                    'INSERT INTO UQ VALUES (?, ?, ?, ?)',
                    (str(date), uq, self.title, self.url)
                    )
                records_in += 1
            except sqlite3.IntegrityError:
                config.LOGGER.info(
                    f'{date} (UQ: {uq}) was found in the DB! Skipped.'
                    )
                continue

        if self.is_url:
            config.DB.commit()
            config.LOGGER.info(f'Wrote {records_in} records into database.')
        else:
            print('Example results:', self.schedule)


class MainPage:
    """Represents the main news page for Urgent Quests."""

    URL = 'https://pso2.com/news/urgent-quests'

    def __init__(self, is_url: bool = True) -> None:
        """Initialize main page for scraping.

        Args:
            is_url (bool, optional): are we using the real URL?
                defaults to None

        """
        config.LOGGER.info('Initializing UQ MainPage...')
        self.is_url = is_url
        if is_url:
            page = requests.get(self.URL)
            self.soup = BeautifulSoup(page.text, 'html.parser')
        else:
            with open(EXAMPLE_MAIN, 'r') as example:
                self.soup = BeautifulSoup(example, 'html.parser')
        try:
            self.schedules = {result[2]: result[3] for result in config.RESULTS}
        except TypeError:
            self.schedules = None

    def parse(self) -> None:
        """Parse the page to find individual schedules."""
        self.new_schedules = {}
        news = self.soup.find('div', 'all-news-section')
        for schedule in news.find_all('div', 'content'):
            title = schedule.find('h3', 'title').text
            link = schedule.find('a', 'read-more')
            if not self.is_url:
                print('Example schedule title:', title)
                continue
            sched_link = link['onclick'].split("'")[1]
            url = f'{self.URL}/{sched_link}'
            if url in config.UQ_BLACKLIST:
                config.LOGGER.info('Found a blacklisted schedule; skipped.')
                config.LOGGER.info(f'- Match title: {title}')
                config.LOGGER.info(f'- Match URL:   {url}')
                continue
            self.new_schedules[title] = url
            if title in self.schedules and url == self.schedules[title]:
                config.LOGGER.info('Found a matching schedule; skipped.')
                config.LOGGER.info(f'- Match title: {title}')
                config.LOGGER.info(f'- Match URL:   {url}')
            else:
                sleep(10)
                s = Schedule(url, title=title)
                s.parse()

        if self.is_url:
            urls = self.new_schedules.values()
            for schedule, url in self.schedules.items():
                if url not in urls:
                    config.LOGGER.info('Deleting records from the following:')
                    config.LOGGER.info(f'- Title: {schedule}')
                    config.LOGGER.info(f'- URL: {url}')
                    config.CURSOR.execute(
                        'DELETE FROM UQ WHERE TITLE = ? AND URL = ?',
                        (schedule, url)
                        )
            config.DB.commit()


if __name__ == '__main__':
    for schedule in EXAMPLE_SCHEDS:
        s = Schedule(schedule, is_url=False)
        s.parse()
        sleep(2)
    mp = MainPage(is_url=False)
    mp.parse()
